helper/keysutil/encrypted_key_storage.go

```diff
@@ -0,0 +1,267 @@
+
+	for i, k := range keys {
+		raw, ok := s.lru.Get(k)
+		if ok {
+			// cache HIT, we can bail early and skip the decode & decrypt operations.
+			decryptedKeys[i] = raw.(string)
+			continue
+		}
+
+		// If a folder is included in the keys it will have a trailing "/".
+		// We need to remove this before decoding/decrypting and add it back
+		// later.
+		appendSlash := strings.HasSuffix(k, "/")
+		if appendSlash {
+			k = strings.TrimSuffix(k, "/")
+		}
+
+		decoded := Base58Decode(k)
+		if len(decoded) == 0 {
+			return nil, errors.New("Could not decode key")
+		}
+
+		// Decrypt the data with the object's key policy.
+		encodedPlaintext, err := s.policy.Decrypt(context, nil, string(decoded[:]))
+		if err != nil {
+			return nil, err
+		}
+
+		// The plaintext is still base64 encoded, decode it.
+		decoded, err = base64.StdEncoding.DecodeString(encodedPlaintext)
+		if err != nil {
+			return nil, err
+		}
+
+		plaintext := string(decoded[:])
+
+		// Add the slash back to the plaintext value
+		if appendSlash {
+			plaintext += "/"
+			k += "/"
+		}
+
+		// We want to store the unencoded version of the key in the cache.
+		// This will make it more performent when it's a HIT.
+		s.lru.Add(k, plaintext)
+
+		decryptedKeys[i] = plaintext
+	}
+
+	return decryptedKeys, nil
+}
+
+// Get implements the logical.Storage Get method.
+func (s *EncryptedKeyStorage) Get(ctx context.Context, path string) (*logical.StorageEntry, error) {
+	encPath, err := s.encryptPath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.s.Get(ctx, encPath)
+}
+
+// Put implements the logical.Storage Put method.
+func (s *EncryptedKeyStorage) Put(ctx context.Context, entry *logical.StorageEntry) error {
+	encPath, err := s.encryptPath(entry.Key)
+	if err != nil {
+		return err
+	}
+	e := &logical.StorageEntry{}
+	*e = *entry
+
+	e.Key = encPath
+
+	return s.s.Put(ctx, e)
+}
+
+// Delete implements the logical.Storage Delete method.
+func (s *EncryptedKeyStorage) Delete(ctx context.Context, path string) error {
+	encPath, err := s.encryptPath(path)
+	if err != nil {
+		return err
+	}
+
+	return s.s.Delete(ctx, encPath)
+}
+
+// encryptPath takes a plaintext path and encrypts each path section (separated
+// by "/") with the object's key policy. The context for each encryption is the
+// plaintext path prefix for the key.
+func (s *EncryptedKeyStorage) encryptPath(path string) (string, error) {
+	path = paths.Clean(path)
+
+	// Trim the prefix if it starts with a "/"
+	path = strings.TrimPrefix(path, "/")
+
+	parts := strings.Split(path, "/")
+
+	encPath := s.prefix
+	context := s.prefix
+	for _, p := range parts {
+		encoded := base64.StdEncoding.EncodeToString([]byte(p))
+		ciphertext, err := s.policy.Encrypt(0, []byte(context), nil, encoded)
+		if err != nil {
+			return "", err
+		}
+
+		encPath = paths.Join(encPath, Base58Encode([]byte(ciphertext)))
+		context = paths.Join(context, p)
+	}
+
+	return encPath, nil
+}
+
+func Base58Encode(buf []byte) string {
```

Comment:
Now that Go 1.10 is out and we're using big.Int, let's make it Base62Encode because why not!